symbolic/src/Data/Macaw/Symbolic/Memory/Lazy.hs

```diff
@@ -0,0 +1,826 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE EmptyCase #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+-- | This module provides a drop-in replacement for
+-- "Data.Macaw.Symbolic.Memory". Unlike the memory model configuration in that
+-- model, which populates the entire static memory in an SMT array before
+-- beginning simulation, the memory model in this module lazily populates the
+-- SMT array. That is, it will begin simulation without populating the SMT array
+-- at all, and it instead checks before each read and write to see if a certain
+-- part of the SMT array needs to be populated. This memory model also makes an
+-- effort to avoid reading from the SMT array if performing concrete reads from
+-- read-only sections of memory. For the full details on how this works, see
+-- @Note [Lazy memory model]@.
+--
+-- Because the API in this module clashes with the API in
+-- "Data.Macaw.Symbolic.Memory", it is recommended that you import these modules
+-- with qualified imports.
+module Data.Macaw.Symbolic.Memory.Lazy (
+  -- * Memory Management
+  memModelConfig,
+  MemPtrTable,
+  MSMC.toCrucibleEndian,
+  MSMC.fromCrucibleEndian,
+  MSMC.GlobalMemoryHooks(..),
+  MSMC.defaultGlobalMemoryHooks,
+  newMergedGlobalMemoryWith,
```

Comment:
Note to self: we should also offer versions of `newGlobalMemory` and `newGlobalMemoryWith` for symmetry with `Data.Macaw.Symbolic.Memory`'s API.